### PR TITLE
Maintain manifoldness when converting between PolySet and ManifoldGeometry

### LINKED ISSUE #5022: Some Manifold objects fail to import into Manifold
When importing files into Manifold, we first convert to a polygon soup, and then try to repair it.
This doesn't work for certain types of manifolds, specifically meshes which touch itself in a single face:

https://github.com/openscad/openscad/blob/98c132516a80312604c346557af52fbe7091638f/tests/data/scad/3D/misc/polyhedron-self-touch-face.scad

To fix this, it would be interesting to try to bypass the polygon soup step in cases where the imported object is a valid manifold.
This also happens with `polyhedron()` definitions, not just imports.

### PR DESCRIPTION
**Pull request by kintel**
Fixes #5022

* createManifoldFromPolySet() is now the main entry point for obtaining a ManifoldGeometry from a PolySet
* We first try direct creation from vertices and indices (after triangulating if necessary)
* Added `ResultObject::mutableResult()` and `ResultObject::constResult()` to make it more clear what to expect, plus clean up some wrong constness.
* Fixed a test case with wrong polygon winding: tests/data/scad/stl/stl-export.scad

### CODE REVIEW

**Comment by kintel** [comment]
@pca006132 This is not mergeable, but could be a starting point for experimentation/discussion.

**Comment by pca006132** [comment]
Wow this is a pretty large patch, I think we can probably start small and try to construct `ManifoldGeometry` from some primitives and `polyhedron` calls?

**Comment by kintel** [comment]
Oh, don't worry, just look at the very last commit.

**Comment by pca006132** [comment]
Seems fine, not sure how the status thing will interact with multiple threads though (not that it matters now). I think manifold should probably provide some thread safe alternative to check the status, though it may be a hard problem (to keep compatibility).

**Comment by kintel** [comment]
What part of the status handling is not thread safe?

**Comment by kintel** [comment]
@pca006132 If you find the time: The interesting pieces are `createManifoldFromPolySet()` and `createManifoldFromTriangularPolySet()`

File: src/geometry/manifold/manifoldutils.cc [hunk 1]
```diff
@@ -102,49 +70,94 @@ std::shared_ptr<const ManifoldGeometry> createManifoldFromSurfaceMesh(const Tria
...
+  // 1. If the PolySet is already manifold, we should be able to build a Manifold object directly
+  // (through using manifold::Mesh).
+  // We need to make sure our PolySet is triangulated before doing that.
+  // Note: We currently don't have a way of directly checking if a PolySet is manifold,
+  // so we just try converting to a Manifold object and check its status. 
+  std::unique_ptr<const PolySet> triangulated;
+  if (!ps.isTriangular()) {
+    triangulated = PolySetUtils::tessellate_faces(ps);
+  }
+  const PolySet triangle_set = ps.isTriangular() ? ps : *triangulated;
```
[the remark this answers is not in the document]
Comment (kintel) [reply src/geometry/manifold/manifoldutils.cc:108]: Good catch! I was not intending to copy.

**Pull request merged**